core-processor/src/processor.rs

```diff
@@ -40,33 +48,201 @@ enum SuccessfulDispatchResultKind {
+    Error(Vec<JournalNote>),
+}
+
+/// Prepares environment for the execution of a program.
+/// Checks if there is a required export and tries to pre-charge for memory pages.
+/// Returns either `PreparedMessageExecutionContext` for `process` or an array of journal notes.
+/// See `PrepareResult` for details.
+pub fn prepare(
     block_config: &BlockConfig,
     execution_context: MessageExecutionContext,
-) -> Vec<JournalNote> {
+) -> PrepareResult {
     let MessageExecutionContext {
         actor,
         dispatch,
         origin,
         gas_allowance,
+        subsequent_execution,
     } = execution_context;
     let Actor {
         balance,
         destination_program,
         executable_data,
     } = actor;
 
-    match check_is_executable(executable_data, &dispatch) {
-        Err(exit_code) => process_non_executable(dispatch, destination_program, exit_code),
-        Ok(data) => process_executable::<A, E>(
-            origin,
-            gas_allowance,
-            data,
+    let (program, pages_with_data) = match check_is_executable(executable_data, &dispatch) {
+        Err(exit_code) => {
+            return PrepareResult::Error(process_non_executable(
+                dispatch,
+                destination_program,
+                exit_code,
+            ))
+        }
+        Ok(ExecutableActorData {
+            program,
+            pages_with_data,
+        }) => (program, pages_with_data),
+    };
+
+    let program_id = program.id();
+    let mut gas_counter = GasCounter::new(dispatch.gas_limit());
+    if !program.code().exports().contains(&dispatch.kind()) {
+        return PrepareResult::WontExecute(process_success(
+            SuccessfulDispatchResultKind::Success,
+            DispatchResult::success(dispatch, program_id, gas_counter.into()),
+        ));
+    }
+
+    let mut gas_allowance_counter = GasAllowanceCounter::new(gas_allowance);
+    let memory_size = match executor::charge_gas_for_pages(
+        &block_config.allocations_config,
+        &mut gas_counter,
+        &mut gas_allowance_counter,
+        program.get_allocations(),
+        program.static_pages(),
+        dispatch.context().is_none() && matches!(dispatch.kind(), DispatchKind::Init),
+        subsequent_execution,
+    ) {
+        Ok(size) => {
+            log::debug!("Charged for memory pages. Size: {size:?}");
+            size
+        }
+        Err(reason) => {
+            log::debug!("Failed to charge for memory pages: {reason:?}");
+            return PrepareResult::Error(match reason {
+                ExecutionErrorReason::InitialMemoryBlockGasExceeded
+                | ExecutionErrorReason::GrowMemoryBlockGasExceeded
+                | ExecutionErrorReason::LoadMemoryBlockGasExceeded => {
+                    process_allowance_exceed(dispatch, program_id, gas_counter.burned())
+                }
+                _ => process_error(dispatch, program_id, gas_counter.burned(), reason),
```

Comment:
is it possible case?

Reply:
Yep, it is possible. `charge_gas_for_pages` may return `LoadMemoryGasExceede`/`GrowMemoryGasExceeded` and `InitialMemoryGasExceeded`.

Reply:
Oh, I got confused by pretty similar enum variants 😅